### PR TITLE
CI: Check for timeout before assertFilesEqual()

### PR DESCRIPTION
**Pull request by andrewbird**
If we are diffing a DOS program's output file against a reference, we should capture the normal output and check that it didn't timeout before completing the output file.

### CODE REVIEW

**Comment by stsp** [comment]
Good, timeouts now seen.
So only fprem FPU test timeouts.
Could you please add the timeout
increment patch to this PR to get
things passed?
I'll likely make that functionality
optional anyway, but we need to
make sure things are now reliable,
and for how much the time have
degraded.

**Comment by stsp** [comment]
But no way to increase just
fprem test?

**Comment by stsp** [comment]
Its not slow comcom, its slow
built-in ANSI. :)

**Comment by andrewbird** [comment]
Not at present, but give me a few minutes.

**Comment by andrewbird** [comment]
Actually it already was just the fprem tests that got increased timeout, purely because they work differently to the other FPU tests and require diffing an output file to a reference file. Other FPU tests work like a lot of other tests, where the test binaries emit PASS or FAIL and we just test for that keyword in the output.

**Comment by stsp** [comment]
Note that fprem already timeouted.

**Comment by stsp** [comment]
I opened the report about this problem:
https://github.com/jwt27/djgpp-cvs/issues/20

**Comment by andrewbird** [comment]
I increased the timeout to your original value (20 mins)

**Comment by stsp** [comment]
Currently building on LP the
version that runs tests w/o
slowdown. Of course its not
the proper optimization, but
at least tests will work again.

**Comment by stsp** [comment]
Was it a timeout?
I think not.
Currently building a fix here:
https://code.launchpad.net/~dosemu2/+recipe/dj64dev-daily-1

**Comment by andrewbird** [comment]
Didn't see a timeout, but thought I'd try to eliminate that first!

**Comment by andrewbird** [comment]
No, still failing. I'll drop the timeout increase when you've fixed everything.

Incidentally it seems the necessary timeout increase for fprem was for native dpmi.
~~~
Test PP-DOS-GIT  FPU fprem JIT vm86 + JIT DPMI                                   ... ok ( 12.94s)
Test PP-DOS-GIT  FPU fprem JIT vm86 + native DPMI                                ... ok (377.91s)
Test PP-DOS-GIT  FPU fprem simulated vm86 + native DPMI                          ... ok (379.05s)
~~~

**Comment by stsp** [comment]
Now test_fat12_img_d_writable fails,
seemingly because of extra CRs.
What I don't understand is why an
xpt log file have extra CRs by itself?
I compare it to the plain output where
there are no extra CRs. Can it be
that you convert lineendings somewhere
in pythos?

**Comment by andrewbird** [comment]
I'll have a look.

**Comment by andrewbird** [comment]
I added a commit that strips the CRs from the output. However it still fails because the written file size is wrong. regex is looking for 8 bytes, but the file is only 7.

File: test/common_framework.py [hunk 1]
```diff
@@ -662,8 +662,18 @@ def runDosemu(self, cmd, opts=None, outfile=None, config=DOSEMU_CONF_DEFAULT, ti
         child = pexpect.spawn(dbin, args)
         ret = ''
         with open(self.logfiles['xpt'][0], "wb") as fout:
-            child.logfile = fout
-            child.setecho(False)
+
+            # Let's strip the CRs from the stream as we write
```
Comment (stsp) [inline test/common_framework.py]: No, why is this?
You need to NOT ADD them, neither
strip. The log must be exactly as it was,
as any permutation only troubles the
debugging process.

Comment (andrewbird) [reply test/common_framework.py]: I don't add them, I get the stream from pexpect which is capturing it from a pty, over which I have no control.

Comment (stsp) [reply test/common_framework.py]: Then likely you need to setup
pty with termios. There are those
`ONLCR | OCRNL` flags, so yes,
you have the full control.

Comment (andrewbird) [reply test/common_framework.py]: I thought I'd posted earlier, **but** the failure is not caused by white space in the CI. Instead the file size is wrong on the FAT. All the other DOSes write 8 bytes, FDPP/Comcom now writes 7. To prove this I just used `wde` on the D: drive, the results:
`68 65 6c 6c 6f 20 0d 0a`  MS-DOS 6.22
`68 65 6c 6c 6f 20 0a`       FDPP/Comcom

Comment (stsp) [reply test/common_framework.py]: Yes, I realize that.
As soon as LP is able to build
something, I'll get the solution
ready. But there is enough time
for you to fix CI, as LP doesn't
work well.

Comment (stsp) [reply test/common_framework.py]: Don't worry, I'll just apply this:
```
+        attrs = termios.tcgetattr(child.fileno())
+        # Disable the ONLCR flag (prevents \n -> \r\n translation)
+        attrs[1] &= ~termios.ONLCR
+        termios.tcsetattr(child.fileno(), termios.TCSANOW, attrs)
```
(AI)

Comment (andrewbird) [reply test/common_framework.py]: Okay will try things out against MS-DOS 6.22 for now.

This looks like the best I can do with terminal settings
```diff
ajb@calypso:/clients/common/dosemu2.git$ git diff
diff --git a/test/common_framework.py b/test/common_framework.py
index e471fcd9d..d3c0c102d 100644
--- a/test/common_framework.py
+++ b/test/common_framework.py
@@ -1,5 +1,6 @@
 import inspect
 import pexpect
+import termios
 import string
 import random
 import re
@@ -660,6 +661,13 @@ class BaseTestCase(object):
             _exit(0)  # Don't let unittest handle it, just exit
 
         child = pexpect.spawn(dbin, args)
+
+        # Tweak the pty to eliminate double CRs
+        fd = child.fileno()
+        attrs = termios.tcgetattr(fd)
+        attrs[1] &= ~termios.ONLCR
+        termios.tcsetattr(fd, termios.TCSANOW, attrs)
+
```

~~~
C:\>testit.bat
testit.bat^M^M
^M
C:\>D:^M
^M
D:\>mkdir test^M
^M
D:\>echo hello > hello.txt^M
^M
D:\>DIR^M
^M
 Volume in drive D has no label^M
 Volume Serial Number is 6A63-9E32^M
 Directory of D:\^M
^M
THERE    TXT             5 07-21-26   8:21p^M
TEST         <DIR>         07-21-26   8:21p^M
HELLO    TXT             8 07-21-26   8:21p^M
        3 file(s)             13 bytes^M
                      21,233,664 bytes free^M
^M
D:\>rem end^M
~~~

Comment (stsp) [reply test/common_framework.py]: Which is exactly what I showed
above from the local commit.
Doesn't matter though, if you add
your commit, I'll remove mine.
comcom64 is still building...

Comment (andrewbird) [reply test/common_framework.py]: A half finished post, and went for dinner, came back and hit send. Now I see what you've written.

Comment (andrewbird) [reply test/common_framework.py]: So would you want further translation by processing of input stream in CI (like the last attempt I pushed) or are you happy with the .xpt file having `\r\n` line endings?

Comment (stsp) [reply test/common_framework.py]: I need the exact match to the original,
and no post-processing.

Comment (andrewbird) [reply test/common_framework.py]: Cool. Will just run full test here first, then push.

**Comment by andrewbird** [comment]
just pushed, as ran okay on MS-DOS 6.22.

**Comment by andrewbird** [comment]
Odd failure with libi86 test 5,, as far as I know dosemu isn't even involved on test. 5. Also the same test passed with msdos..

**Comment by andrewbird** [comment]
We could probably drop the fprem timeout increase commit now, as the longest fprem time in the test was 23 secs?

**Comment by stsp** [comment]
Where do you see failure?
It all passed.
Remove timeout if you want.

**Comment by stsp** [comment]
https://github.com/dosemu2/dosemu2/actions/runs/29709227056/job/88763411663
Can you look into this run,
is it a timeouted log again,
or something else? The log
is clearly incomplete.

**Comment by andrewbird** [comment]
Okay, I'm going to try adding the assert to the runDosemu command itself, and remove it from all the callers. At the moment I can't think why we'd ever not want to assert if we got a timeout, but I it's possible I've overlooked a corner case. Will run all the tests locally that normally get run on the scheduled (but not ASAN, or packages), to see before I push it here.

**Comment by andrewbird** [comment]
> Where do you see failure?

I think it was from my repo Actions from when I pushed to the branch. It even may have been an older version that ran for a long while before eventually failing. Any I'll ignore for now.

**Comment by stsp** [comment]
> Okay, I'm going to try adding the assert to the runDosemu command itself, and remove it from all the callers.

But do you see what exactly
caller currently misses it?

**Comment by andrewbird** [comment]
The test reported it as a parse error, but you can clearly see it's a timeout. So we need to apply that check to all the tests I think. However the test_dosemu.py had been running for 2220 secs and the DOS output was double spaced. Also the test.processor.py took 1957 secs before it, so together that's ~1.5 hours  before the report was generated. So I think it's an older run from before you fixed comcom which somehow survived up the point that one test decided it had taken too long.

**Comment by stsp** [comment]
> but you can clearly see it's a timeout.

How?
Is there a word "timeout" somewhere?
I am just not entirely sure, hence the
questions.

**Comment by andrewbird** [comment]
~~~
Traceback (most recent call last):
  File "/home/runner/work/dosemu2/dosemu2/test/test_dosemu.py", line 995, in test_memory_ems_borland
    memory_ems_borland(self)
  File "/home/runner/work/dosemu2/dosemu2/test/func_memory_ems_borland.py", line 31, in memory_ems_borland
    raise self.failureException("Parse Error:\n" + results) from None
AssertionError: Parse Error:
Timeout: 40 seconds waiting for ['rem end']
~~~

Last line

**Comment by stsp** [comment]
Okey, I certainly got confused by:
`    raise self.failureException("Parse Error:\n" + results) from None`
as what kind of None?

`AssertionError: Parse Error:`
And an empty string after "Parse Error:"...
All that made me believe there is no
data to parse, so I didn't look at a line
below.
Maybe it could print something more
than empty line after "Parse Error", so
to see what it parsed and what's an
error?

**Comment by andrewbird** [comment]
It's a Pythonism, if you're already handling an exception and you raise another (say self.failureException) you'll get an error saying attempting to handle an exception whilst already handling one. The fix is that you raise the second 'from None' then there's no nesting.

**Comment by stsp** [comment]
> So I think it's an older run from before you fixed comcom which somehow survived up the point that one
> test decided it had taken too long.

Its not fully fixed even now, I am
only CIing the real "fix" now.
I think for ASAN there should be
no timeouts like 40s, so lets also
increase that one.

**Comment by stsp** [comment]
> second 'from None' then there's no nesting.

But why `Parse Error:` and no info
after colon?

**Comment by andrewbird** [comment]
> All that made me believe there is no
data to parse

Will look into that, not sure if we currently return partial results in the event of timeout. I presume that's desirable then?

**Comment by andrewbird** [comment]
asan currently gets an increased timeout of 45 secs per test which is added to all the timeouts specified and the default value. We could increase that, but we need to ensure it's not too high or Github will kill the whole workflow and we'll get no logs at all in that case.

**Comment by stsp** [comment]
> Will look into that, not sure if we currently return partial results in the event of timeout. I presume that's
> desirable then?

In case of timeout, the parse error
should not happen indeed. But as
long as it currently does, you can
fix it first, and then add timeout
handling so that parse error doesn't
happen at all. :)

**Comment by stsp** [comment]
> asan currently gets an increased timeout of 45 secs per test

And what I see, is:
```
Timeout: 40 seconds waiting for ['rem end']
```
So where is the added 45?

**Comment by andrewbird** [comment]
Ahh, a flaw. The default timeout is added to the batch file or executable we are running, not the boilerplate around it. I'll fix that too.

**Comment by andrewbird** [comment]
I dropped one commit that doesn't belong in this PR series, other than that it should be good to apply. Assuming of course there's nothing further?

**Comment by stsp** [comment]
If asan passes, that's already great.
Btw. how much could we save by
marking libi86 tests as expensive?
Waiting 2hrs is a bit annoying, although
of course LP forces the much larger
waits.

**Comment by andrewbird** [comment]
They already are marked as such, however the expensive switch is only used on the 'simple' run which at the moment is only

- scheduled / asan
- triggered / when it's a push to dosemu-devel

**Comment by stsp** [comment]
You mean, a push to PR branch
doesn't run those?

**Comment by andrewbird** [comment]
Actually I don't think we've had a run with asan on this branch.  Not sure if I can run a manual asan run on mine. If I can do it my changing by default branch to it, that might work.

**Comment by andrewbird** [comment]
I mean a push to any topic branch (including a PR branch) runs 'normal', which is PPDOSGit + MS-DOS 6.22 and all tests except those tagged a 'unreliable' i.e. PIT

**Comment by andrewbird** [comment]
Oh, asan is only run as part of the scheduled weekly test, it just takes too long. You'd be getting failures from a push to a topic branch 3 hours ago, and in the meantime it's been pushed to four times since.

**Comment by andrewbird** [comment]
I'm running a manual asan/normal test now on my ci-267 topic branch (temporarily my default)

**Comment by stsp** [comment]
Thanks!
Lets see if asan test now works.
Another heroic CI recovery efforts.
I wonder for how long will it last.

**Comment by stsp** [comment]
Ah, asan test will fail again, unless
I run it again, rather than restart?
Ok, won't bother then. :)

**Comment by andrewbird** [comment]
Mmm, looking at my manual asan test, there's a reason that I don't run asan with anything but simple, it just takes so much longer per test (8 second boot for instance). It seems that even with triple the normal test time for freedos kernel build it's still not enough. https://github.com/andrewbird/dosemu2/actions/runs/29943200916/job/89002011565 I'll let this run, but probably we'll hit the github total job timeout and lose the logs.
I'll run another but asan-simple, so we can see if it's clean.

**Comment by andrewbird** [comment]
Here's the asan-simple which completed without error https://github.com/andrewbird/dosemu2/actions/runs/29950081771/job/89025315961

**Comment by stsp** [comment]
Good to know, thanks.
This probably means that this
week I won't get a failure from
weekly test. "no failure emails"
periods are very infrequent.
